iceoryx_examples/automotive_soa/include/owl/kom/event_subscriber.inl

```diff
@@ -0,0 +1,163 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_SUBSCRIBER_INL
+#define IOX_EXAMPLES_AUTOMOTIVE_SOA_EVENT_SUBSCRIBER_INL
+
+#include "owl/kom/event_subscriber.hpp"
+
+namespace owl
+{
+namespace kom
+{
+template <typename T>
+inline EventSubscriber<T>::EventSubscriber(const ServiceIdentifier& service,
+                                           const InstanceIdentifier& instance,
+                                           const EventIdentifier& event) noexcept
+    : m_serviceDescription(service, instance, event)
+{
+}
+
+template <typename T>
+inline void EventSubscriber<T>::Subscribe(std::size_t queueCapacity) noexcept
+{
+    if (HasReceiveCallback())
+    {
+        UnsetReceiveCallback();
+    }
+
+    m_subscriber.emplace(
+        m_serviceDescription,
+        iox::popo::SubscriberOptions{queueCapacity, HISTORY_REQUEST, iox::NodeName_t(), NOT_OFFERED_ON_CREATE});
+    m_subscriber.value().subscribe();
+}
+
+template <typename T>
+inline void EventSubscriber<T>::Unsubscribe() noexcept
+{
+    if (!m_subscriber.has_value())
+    {
+        return;
+    }
+
+    if (HasReceiveCallback())
+    {
+        UnsetReceiveCallback();
+    }
+
+    m_subscriber.value().unsubscribe();
+    m_subscriber.reset();
+}
+
+template <typename T>
+template <typename Callable>
+inline core::Result<size_t> EventSubscriber<T>::TakeNewSamples(Callable&& callable, size_t maxNumberOfSamples) noexcept
+{
+    core::Result<size_t> numberOfSamples{0};
+
+    if (!m_subscriber.has_value())
+    {
+        return numberOfSamples;
+    }
+
+    // Hint: Depending on the type of callable, it needs to be checked for null or restricting to lambdas
+
+    while (m_subscriber.value()
+               .take()
+               .and_then([&](const auto& sample) {
+                   callable(sample.get());
+                   numberOfSamples++;
```

Comment:
Why not `++numberOfSamples` the post increment comes in theory with overhead and also provides the wrong intention to fellow developers

Reply:
If the type is not a fundamental type and the result is used in the same expression statement a pre-increment might be more efficient some CPU architecture with some compilers.

However, if the type is a foundamental type and the result is not used in the same expression statement, pre- and post-increment do not differ performance-wise.

What kind of intention do you read here? I read "`numberOfSamples` is incremented so it can be read out and returned at a later point".